_posts/2022-12-12-deltalake-spark-testing-learnings.md

```diff
@@ -0,0 +1,234 @@
+
+### Writing Test
+
+If you have come this far, we have one last task at hand. That is adapting our notebook code to be testable. The easiest way to show this is via before and after comparison of what a typical structure would be vs post refactoring
+
+So lets say we have a hypothetical scenario where we want to test a transformation function we have written that takes a dataframe containing a list of people and we want to filter out where their date of birth is equal to or greater than 01 Jan 1991, we can structure our code as below.
+
+
+*databricks_notebook.py*
+```
+import pyspark.sql.functions as F
+from os import env
+
+def transform_dataset(df):
+    return df.select("name", "dob").where(F.col("dob") >= F.lit("1999-01-01") )
+
+def run_main_loop():
+    # run main job here
+
+try:
+    is_live = dbutils.widgets.get("is_live").lower() == "true"
+
+except Exception as e:
+    print(f'There was an issue initialising parameters. {e}')
+    is_live = False
+
+if is_live == True:
+    run_main_loop()
+```
+
+It firstly attempts to use Databrick's global dbutils to get configuration parameters. If that succeeds then we run the main loop, else it would throw an exception of undefined variable access and we just treat this notebook as offline mode so to speak.
+
+So when we get to a stage of writing our unit test, we can import this file as such:
+
+*databricks_notebook_test.py*
+```
+import pyspark.sql.types as T
+from databricks_notebook import transform_dataset
+import pyspark.sql.functions as F
+from datetime import date
+
+def transform_dataset(df):
+    return df.select("name", "dob").where(F.col("dob") >= F.lit("1999-01-01") )
+
+def test_date_filter(spark):
+    schema = T.StructType([
+        T.StructField("name", T.StringType(), True),
+        T.StructField("dob", T.DateType(), True),
+    ])
+
+    data = [
+    (
+        "John",
+        date(1978, 5, 8)
+    ),
+    (
+        "Tim",
+        date(1999, 2, 8)
+    ),
+    (
+        "James",
+       date(1996, 6, 9)
+    )]
+    
+    df = spark.createDataFrame(data=data, schema=schema)
+    transformed_df = transform_dataset(df)
+
+    results = transformed_df.collect()
+    
+    assert len(results) == 1
+    assert results[0]["name"] == "Tim"
+   
+
+```
+
+### Troubleshooting
+
+Q: I've got the latest version of Python but I am receiving error such as `_pickle.PicklingError: Could not serialize object: IndexError: tuple index out of range` when I run my tests... halp!
+
+A: Check the Py readiness docs https://pyreadiness.org/3.11/ to see that PySpark is supported for the desired python version (i.e. 3.11).
+
+
+### Conclusion
+
+Well thats about it!, hope this was helpful by some measure. 
+
+As a final note, you can expand the scope of your test if you wish by importing the code that creates your destination table for the initial setup, then running the function that transforms and merges the data in there. 
```

Comment:
Perhaps a final call to action? Contact us for any databricks projects (with a link)

Reply:
phew, got that in